cpp/src/streaming/cudf/shuffler.cpp

```diff
@@ -104,33 +102,18 @@ coro::task<std::vector<PackedData>> ShufflerAsync::extract_async(shuffler::PartI
     );
 
     co_await cv_.wait(lock, [this, pid]() {
-        // Note: purposefully not checking for extracted_pids_.contains(pid) here,
-        // because it would require notifying the cv every time a partition is extracted.
-        // Consequence of this is that, if pid was extracted by some other task, this task
-        // would only be notified during the all_extracted check.
         return all_extracted_unsafe() || ready_pids_.contains(pid);
     });
 
-    // partition not found (may have been already extracted or shuffler was finished
-    // before the pid was inserted into ready_pids_)
-    RAPIDSMPF_EXPECTS(
-        !extracted_pids_.contains(pid) && ready_pids_.erase(pid) > 0,
-        "partition already extracted or not found: " + std::to_string(pid),
-        std::out_of_range
-    );
-    extracted_pids_.emplace(pid);
-    auto all_extracted = all_extracted_unsafe();
-
-    auto chunks = shuffler_.extract(pid);
-    lock.unlock();
-
-    // if all partitions have been extracted, notify all waiting tasks.
-    if (all_extracted) {
-        ctx_->comm()->logger().trace("all partitions extracted");
-        co_await cv_.notify_all();
+    // Did we wake up because all partitions have been extracted?.
+    if (all_extracted_unsafe()) {
+        co_return std::nullopt;
     }
 
-    co_return std::move(chunks);
+    // pid has now been extracted and isn't ready anymore.
+    RAPIDSMPF_EXPECTS(ready_pids_.erase(pid) > 0, "something went wrong");
+    RAPIDSMPF_EXPECTS(extracted_pids_.emplace(pid).second, "something went wrong");
+    co_return shuffler_.extract(pid);
```

Comment:
I think we need to notify `cv_`, if this is the last PID. Otherwise we'd have to rely on spurious wake-ups in some edge cases.

Reply:
But we notify **all**. So if we have many consumers waiting, all of them wakes up

Reply:
Consumers (extractors) could miss the notification from the callback. Lets take 1 partition 2 consumers again. 
Say, by the time extract any is scheduled, the shuffle has completed message passing and the callbacks have already been called. Then consumers miss the cv notification. One consumer will bypass cv wait and consume the partition, but the other would still be waiting on the cv.

Reply:
we only suspend if the predicate is not satisfied. in your example I think the predicate is satisfied so we won't suspend, no?

specifically the predicate is called with the lock held so if two people come in for one partition one of them gets the lock extracts the partition with the lock then unlocks which wakes the next person who sees that the predicate is true and returns because everything has been extracted

Reply:
Ah! good point @wence- . You are absolutely right. If we had two extract_any consumers for a single partition, it is true. 

Okay, let me see if I could counter. 

Say we have many consumers, and 2 partitions (0, 1). consumer 1 extracts PID 0, while other consumers  extracts any. Say, an other consumer locks the mutex first and extracts PID 0. Now, when consumer 1 locks the mutex, it would go into waiting, because not all partitions have been extracted yet, and PID 0 is not available. Later, some other consumer extracts PID 1. But consumer 1 will still be waiting, because no one has is notifying the cv, isnt it?

Reply:
Good catch, I have update the wait in `extract_async()!:
```c++
    co_await cv_.wait(lock, [this, pid]() {
        return ready_pids_.contains(pid) || extracted_pids_.contains(pid);
    });
```
A partition ID is always in exactly one of the two sets:
     - `ready_pids_`: partitions ready for extraction but not yet extracted.
     - `extracted_pids_`: partitions that have already been extracted.

Agree?